### DISCUSSION TITLE
init empty instance

### DISCUSSION THREAD

**Discussion by Ginko2501** [mitsuba-renderer/mitsuba3#732] [Q&A] [answered]
Hi,

I was writing my custom renderer and noticed this subtle difference--

If I initialize my `mi.SurfaceInteraction3f` by `si = mi.SurfaceInteraction3f()`, I cannot call `si.initialize_sh_frame()` later even though I have assigned `si.sh_frame.n`. On the other hand, if I initialize by `si = dr.zeros(mi.SurfaceInteraction3f)`, I will have no problems.

I wonder what are the differences between these two initialization methods, and which one is the standard way to go. Thanks.

**Comment by njroussel** [comment, marked as answer]
Hi @Ginko2501 

The default constructor will just allocate the memory, but will not initialize it.
For `initialize_sh_frame()` to work, both the `si.sh_frame.n` and `dp_du` fields needs to be initialized. This is why it would crash if you only initialize the former.

As a general rule of thumb, I use `dr.zeros` everytime I allocate a new variable. If you take a look at some of the source code (Python or C++) you'll see that this is quite common.

**Reply by Ginko2501** [reply]
Hi, thanks a lot for the reply! You mentioned a couple of times that I should refer to the source code, so I wonder if you could point me to some core files in the source code. I am developing in the Python API a custom renderer that requires autodiff. Thanks again!

**Reply by njroussel** [reply]
The source code for [`prbvolpath` ](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/prbvolpath.py)and [`prb_reparam`](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/prb_reparam.py) is some of the most complete Python source code we have and covers a lot of what one can do with Mitsuba/Dr.Jit.

The real "core" sources are in C++, I'd typically recommend looking at the [`integrator.cpp`](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/render/integrator.cpp). It's a good place to view the full rendering process. In addition, [`common.py`](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/common.py) holds some of the boilerplate and tricks to make the rendering process easily differentiable.
